src/PluginForm/Bluesnap/HostedPaymentFieldsPaymentMethodAddForm.php

```diff
@@ -0,0 +1,214 @@
+use Drupal\commerce_payment\PluginForm\PaymentMethodAddForm as BasePaymentMethodAddForm;
+
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Adds a credit card payment method form for the HostedPaymentFields gateway.
+ *
+ * @package Drupal\commerce_bluesnap\PluginForm\Bluesnap
+ */
+class HostedPaymentFieldsPaymentMethodAddForm extends BasePaymentMethodAddForm {
+
+  /**
+   * The Bluesnap API helper service.
+   *
+   * @var \Drupal\commerce_bluesnap\ApiService
+   */
+  protected $apiService;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(
+    InlineFormManager $inline_form_manager,
+    RouteMatchInterface $route_match,
+    EntityTypeManagerInterface $entity_type_manager,
+    LoggerInterface $logger,
+    ApiService $api_service
+  ) {
+    parent::__construct(
+      $inline_form_manager,
+      $route_match,
+      $entity_type_manager,
+      $logger
+    );
+
+    $this->apiService = $api_service;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('plugin.manager.commerce_inline_form'),
+      $container->get('current_route_match'),
+      $container->get('entity_type.manager'),
+      $container->get('logger.factory')->get('commerce_payment'),
+      $container->get('commerce_bluesnap.api_service')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildCreditCardForm(
+    array $element,
+    FormStateInterface $form_state
+  ) {
+    // The BlueSnap unique Hosted Payment Fields Token that will be sent in the
+    // drupalSettings to the JS.
+    // First, initialize BlueSnap.
+    if (empty($form_state->getValue('bluesnap_token'))) {
+      $plugin = $this->entity->getPaymentGateway()->getPlugin();
+      $bluesnap_token = $this->apiService->getHostedPaymentFieldsToken($plugin);
+    }
+    else {
+      $bluesnap_token = $form_state->getValue('bluesnap_token');
+    }
+
+    // Alter the form with Bluesnap specific needs.
+    $element['#attributes']['class'][] = 'bluesnap-form';
+    $element['#attached']['library'][] = 'commerce_bluesnap/hosted_payment_fields_form';
+    $element['#attached']['drupalSettings']['commerceBluesnap'] = [
+      'hostedPaymentFields' => [
+        'token' => $bluesnap_token,
+      ],
+    ];
+
+    // Hidden fields which will be populated by the js.
+    $this->hiddenFields($element);
+    // The credit card fields necessary for BlueSnap.
+    $this->ccFields($element);
+
+    // To display validation errors.
+    $element['payment_errors'] = [
+      '#type' => 'markup',
+      '#markup' => '<div id="payment-errors"></div>',
+      '#weight' => -200,
+    ];
+
+    return $element;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitCreditCardForm(
```

Comment:
**CHANGE REQUEST**

*API*

Looking at the parent class this is a `protected` method.